### DISCUSSION TITLE
Strange behaviour of form_errors in two different forms

### DISCUSSION THREAD

**Discussion by sebastian0119** [symfony/symfony#58673] [Q&A] [answered]
Hi there!

I am using constraints to validate form fields. In order to format the error messages in a specific way, I am customizing the form_errors in the following way:

```
{%- block form_errors -%}
    {%- if errors|length > 0 -%}
        {% if compound %}
            {# TODO: set background color back to bg-red-100 when there is clarity about location of error messages #}
            <div class="not-prose bg-green-100 border-l-4 border-red-500 text-red-700 px-4 py-2" role="alert">
            {%- for error in errors -%}
                <p class="{{ error_item_class|default('text-red-700') }}"><b>{% trans %}app.error.check{% endtrans %}:</b> {{ error.message }}</p>
            {%- endfor -%}
            </div>
        {% else %}
            <div class="not-prose bg-red-100 border-l-4 border-red-500 text-red-700 px-4 py-2" role="alert">
            {%- for error in errors -%}
                <p class="{{ error_item_class|default('text-red-700') }}"><b>{% trans %}app.error.check{% endtrans %}:</b> {{ error.message }}</p>
            {%- endfor -%}
            </div>
        {% endif %}
    {%- endif -%}
{%- endblock form_errors -%}
```

For easier identification during testing, compound errors were configured with a green background, and non-compound errors with a red background.

Now, I have two different forms (well, more in fact, but for the sake of this discussion just two): one to register new companies and another one for new charges.

Both are implemented in twig by just

```
{{ form_start(form) }}
    {{ form_widget(form) }}
    <button class="btn btn-primary mt-2">{{ button_label|default('Save') }}</button>
{{ form_end(form) }}
```

When testing the validation rules, the company form shows the error messages right next to the corresponding field (this is the approach I prefer). On the contrary, the charge form displays the error messages at the top of the form.

I know that there are compound and non-compound errors and compound errors are displayed at the top of the form, while non-compound errors are displayed right next to the label. But the error messages from the **charge form** should not be compound errors.

But for some reason, they are because a) they have a green background and b) they are displayed at the top of the form.
But when looking at the output of the debug bar, they are clearly associated to a specific form field. 

Following the different examples:

1. The company form with some error messages, displayed next to each form field / label.

![company_form](https://github.com/user-attachments/assets/2e2b410b-6181-44de-b4f3-fbf4d516bd18)

2. A screenshot of the validation log of the debug bar (for the company form)

![company_violations](https://github.com/user-attachments/assets/ba656c18-ec3c-4094-9b4e-1180efe71f2c)

4. The charge form, with all the error messages at the top of the form. In fact, you can see that the label (mentioned in the error message) refers to the global `Charge` object and **not** to a specific form field.

![charge_form](https://github.com/user-attachments/assets/764b32da-22d5-4459-9570-80844d07fe7c)

5. A screenshot of the validation log of the debug bar (charge form), where each violation is clearly referenced to a specific form field

![charge_new](https://github.com/user-attachments/assets/fc7d545c-02ad-4f3b-8de4-9af7f820eb40)

Does anyone have an idea why this is happening? And what mistake I am doing?

Thanks in advance; any help is appreciated.
Sebastian

**Comment by MatTheCat** [comment]
Hello, your charge form code would help but this probably has to do with the `error_bubbling` and/or `error_mapping` options.

**Comment by sebastian0119** [comment]
Thanks, @MatTheCat for taking the time.

At a certain point, I thought so, too. But I am not doing `error_bubbling` and from the documentation, the default is set to false (unless the form is compound, which is not my case).

Anyway: here comes the `ChargeType.php` code:

```
<?php
namespace App\Form;

use App\Entity\Charge;
use App\Entity\Company;
use App\Entity\CompanyLegaltype;
use App\Entity\Currency;
use App\Entity\Project;
use App\Entity\Subcontract;
use Symfony\Bridge\Doctrine\Form\Type\EntityType;
use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
use Symfony\Component\Form\Extension\Core\Type\CollectionType;
use Symfony\Component\Form\Extension\Core\Type\DateType;
use Symfony\Component\Form\Extension\Core\Type\NumberType;
use Symfony\Component\Form\Extension\Core\Type\TextType;
use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;

class ChargeType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options): void
    {
        $builder
            ->add('CompanyLegaltype', EntityType::class, [
                'class' => CompanyLegaltype::class,
                'choice_label' => 'full',
                'placeholder' => '-- Please select --',
                // Makes sure that a simple select field is shown
                'multiple' => false,
                'expanded' => false,
                'required' => true,
                'disabled' => false,
            ])
            ->add('CompanyIdentification', NumberType::class, [
                'required' => true,
            ])
            ->add('CompanyActivity', NumberType::class, [
                'required' => true,
            ])
            ->add('Date', DateType::class, [
                'widget' => 'single_text',
                'required' => true,
            ])
            ->add('Reply', ChoiceType::class, [
                'placeholder' => '-- Please select --',
                'choices'  => [
                    'Aceptado' => 1,
                    'Rechazado' => 3,
                ],
                'required' => true,
            ])
            ->add('Number', NumberType::class, [
                'required' => true,
            ])
            ->add('Code', TextType::class, [
                'required' => true,
            ])
            ->add('Description', TextType::class, [
                'required' => true,
            ])
            ->add('Related')
            ->add('TaxAmount', NumberType::class, [
                'required' => true,
                'scale' => 5,
                'input' => 'string',
            ])
            ->add('TaxCondition', ChoiceType::class, [
                'placeholder' => '-- Please select --',
                'choices'  => [
                    'Genera crédito IVA' => 1,
                    'Genera crédito parcial' => 2,
                    'Gasto; no genera crédito' => 4,
                    'Proporcionalidad' => 5,
                ],
                'required' => true,
            ])
            ->add('TaxCredit', NumberType::class, [
                'required' => true,
                'scale' => 5,
                'input' => 'string',
            ])
            ->add('TaxExpense', NumberType::class, [
                'required' => true,
                'scale' => 5,
                'input' => 'string',
            ])
            ->add('Amount', NumberType::class, [
                'required' => true,
                'scale' => 5,
                'input' => 'string',
            ])
            ->add('Comment')
            ->add('Company', EntityType::class, [
                'class' => Company::class,
                'choice_label' => 'name',
                'placeholder' => '-- Please select --',
                // Makes sure that a simple select field is shown
                'multiple' => false,
                'expanded' => false,
                'required' => true,
                'disabled' => false,
            ])
            ->add('Project', EntityType::class, [
                'class' => Project::class,
                'choice_label' => 'name',
                'placeholder' => '-- Select company --',
                // Makes sure that a simple select field is shown
                'multiple' => false,
                'expanded' => false,
                'required' => true,
                'disabled' => false,
            ])
            ->add('Subcontract', EntityType::class, [
                'class' => Subcontract::class,
                'choice_label' => 'description',
                'placeholder' => '-- Select project --',
                // Makes sure that a simple select field is shown
                'multiple' => false,
                'expanded' => false,
                'required' => true,
                'disabled' => false,
            ])
            ->add('Currency', EntityType::class, [
                'class' => Currency::class,
                'choice_label' => 'short',
                // Makes sure that a simple select field is shown
                'multiple' => false,
                'expanded' => false,
                'required' => true,
                'disabled' => false,
            ])
        ;
    }

    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefaults([
            'data_class' => Charge::class,
        ]);
    }
}
```

**Reply by MatTheCat** [reply]
Indeed this looks like a classic setup. Are you sure another form theme is not used for this one? If not I’m afraid you’ll have to provide a reproducer.

**Reply by sebastian0119** [reply]
Nope - no other form theme is used.

Anyway, what is the preferred way to provide you with a reproducer? Link to zip file?

**Reply by MatTheCat** [reply]
Zip file or GitHub repository, as you want. Ideally we can run the reproducer using Docker.

**Reply by sebastian0119** [reply]
Here is the reproducer:
https://github.com/sebastian0119/symfony_charge_reproducer

I hope I included all relevant files, composer packages, etc. 

I am running my test environment with the built-in Symfony webserver + a local MariaDB database.

Thanks in advance!

**Reply by MatTheCat** [reply, marked as answer]
Your form fields are not named the same than its data properties (e.g. `CompanyLegaltype` vs `companyLegaltype`), so errors cannot be mapped.

> By default, errors for any property that is not mapped will bubble up to the parent form.
>
> https://symfony.com/doc/current/reference/forms/types/form.html#error-mapping

**Reply by sebastian0119** [reply]
Some things are just too obvious to be seen... Sorry for not noticing it myself and thanks for bothering!

**Reply by MatTheCat** [reply]
No problem; sometimes I learn things and sometimes I don't 😄
